### PR TITLE
Overhaul

### PR DESCRIPTION
**Pull request by mdpye**
What it says on the tin.

Includes migration to v2 server side APIs for typing indicators and room memberships.

Removes `Future` from the public APIs in favour of simple blocking calls which but the caller in control of concurrency. A lightweight callback based alternative will be offered.

### CODE REVIEW

**Comment by mdpye** [comment]
[[5] Refactor Android lib](https://trello.com/c/oS0exAbO/1325-5-refactor-android-lib)

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/cursors/CursorSubscriptionEventParser.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import com.pusher.util.asSuccess
+import com.pusher.util.orElse
+import elements.Error
+import elements.Errors
+
+internal object CursorSubscriptionEventParser : DataParser<CursorSubscriptionEvent> {
+    override fun invoke(body: String): Result<CursorSubscriptionEvent, Error> =
+            body.parseAs<JsonElement>()
+                    .map { it.takeIf { it.isJsonObject }?.asJsonObject }
+                    .flatMap { it.orElse { Errors.other("") } }
```
Comment (vivangkumar) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/cursors/CursorSubscriptionEventParser.kt]: I think we should describe the error here.

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/memberships/MembershipSubscriptionEventParser.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import com.pusher.chatkit.memberships.MembershipSubscriptionEvent.*
+import com.pusher.chatkit.util.asString
+import com.pusher.util.asFailure
+import com.pusher.util.orElse
+
+internal object MembershipSubscriptionEventParser : DataParser<MembershipSubscriptionEvent> {
+    override fun invoke(body: String): Result<MembershipSubscriptionEvent, Error> =
+        body.parseAs<JsonElement>()
+            .map { it.takeIf { it.isJsonObject }?.asJsonObject }
+            .flatMap { it.orElse { Errors.other("") } }
```
Comment (vivangkumar) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/memberships/MembershipSubscriptionEventParser.kt]: More descriptive error here too?

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/presence/PresenceSubscriptionEventParser.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import com.pusher.util.orElse
+import elements.Error
+import elements.Errors
+
+internal object PresenceSubscriptionEventParser : DataParser<PresenceSubscriptionEvent> {
+
+    override fun invoke(body: String): Result<PresenceSubscriptionEvent, Error> =
+            body.parseAs<JsonElement>()
+                    .map { it.takeIf { it.isJsonObject }?.asJsonObject }
+                    .flatMap { it.orElse { Errors.other("") } }
```
Comment (vivangkumar) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/presence/PresenceSubscriptionEventParser.kt]: Likewise here

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomSubscriptionEventParser.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import com.pusher.util.asFailure
+import com.pusher.util.orElse
+import elements.Error
+import elements.Errors
+
+internal object RoomSubscriptionEventParser: DataParser<RoomSubscriptionEvent> {
+    override fun invoke(body: String): Result<RoomSubscriptionEvent, Error> =
+            body.parseAs<JsonElement>()
+                    .map { it.takeIf { it.isJsonObject }?.asJsonObject }
+                    .flatMap { it.orElse { Errors.other("") } }
```
Comment (vivangkumar) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomSubscriptionEventParser.kt]: And here..

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomSubscriptionGroup.kt [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+    private fun scheduleStopTypingEvent(user: User): Boolean {
+        synchronized(typingTimers) {
+            val new = typingTimers[user.id] == null
+            if (!new) {
+                typingTimers[user.id]!!.cancel()
+            }
+
+            typingTimers[user.id] = Futures.schedule {
+                Thread.sleep(1_500)
```
Comment (vivangkumar) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/RoomSubscriptionGroup.kt]: Perhaps we should constantise the `1_500` and name it a bit descriptively.

**Pull request merged**